lib/github.ml

```diff
@@ -87,19 +87,25 @@ let parse_exn headers body =
   | "member" | "create" | "delete" | "release" -> Event (event_notification_of_string body)
   | event -> failwith @@ sprintf "unsupported event : %s" event
 
+type basehead = string * string
+
 type gh_link =
   | Pull_request of repository * int
   | Issue of repository * int
   | Commit of repository * commit_hash
+  | Compare of repository * basehead
 
-let gh_link_re = Re2.create_exn {|^(.*)/(.+)/(.+)/(commit|pull|issues)/([a-z0-9]+)/?$|}
+let gh_link_re = Re2.create_exn {|^(.*)/(.+)/(.+)/(commit|pull|issues|compare)/([a-zA-Z0-9/:\-_.~\^]+)$|}
+let commit_sha_re = Re2.create_exn {|[a-f0-9]{4,40}|}
+let comparer_re = {|([a-zA-Z0-9/:\-_.~\^]+)|}
+let compare_basehead_re = Re2.create_exn (sprintf {|%s([.]{3})%s|} comparer_re comparer_re)
 let gh_org_team_re = Re2.create_exn {|[a-zA-Z0-9\-]+/([a-zA-Z0-9\-]+)|}
 
 (** [gh_link_of_string s] parses a URL string [s] to try to match a supported
     GitHub link type, generating repository endpoints if necessary *)
 let gh_link_of_string url_str =
   let url = Uri.of_string url_str in
-  let path = Uri.path url in
+  let path = Uri.pct_decode (Uri.path url) in
```

Comment:
this is suspicious, why is this needed. If you need to decode path components then should split the url into components first (by slash) and then percent-decode each

Reply:
ie gh_commit_re etc will need to be reworked

Reply:
so I should update the regex for capturing the URL first then decode only the path components that need decoding i.e. `item`?

Reply:
break the path by slash, check that it matches the required scheme e.g. `| _::_::("commit"|"pull"|etc)::x::[] -> ...` then decode `x` and match it with regexp

Reply:
got it thanks!